File: server-e/app/Services/HealthReporter.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;

class HealthReporter
{
    public function report(
        string $status,
        string $component,
        string $reason,
        ?string $message = null
    ): bool {
        $payload = [
            'service' => config('health.service'),
            'status' => $status,
            'component' => $component,
            'reason' => $reason,
            'message' => $message,
            'timestamp' => now()->toISOString(),
        ];

        try {
            $response = Http::timeout(3)
                ->withToken(config('health.al_token'))
                ->post(
                    config('health.al_url') . '/api/v1/health-events',
                    $payload
                );

            if ($response->successful()) {
                Log::info('Health event sent to AL', [
                    'payload' => $payload,
                    'response' => $response->json(),
                ]);

                return true;
            }

            Log::error('AL rejected health event', [
                'status' => $response->status(),
                'response' => $response->body(),
                'payload' => $payload,
            ]);

            return false;

        } catch (\Throwable $exception) {

            Log::error('Unable to communicate with AL', [
                'error' => $exception->getMessage(),
                'payload' => $payload,
            ]);

            return false;
        }
    }

    public function healthy(
        string $component,
        string $reason = 'SERVICE_AVAILABLE'
    ): bool {
        return $this->report(
            'HEALTHY',
            $component,
            $reason
        );
    }

    public function degraded(
        string $component,
        string $reason,
        ?string $message = null
    ): bool {
        return $this->report(
            'DEGRADED',
            $component,
            $reason,
            $message
        );
    }

    public function down(
        string $component,
        string $reason,
        ?string $message = null
    ): bool {
        return $this->report(
            'DOWN',
            $component,
            $reason,
            $message
        );
    }

    public function heartbeat(): bool
{
    return $this->report(
        'HEALTHY',
        'application',
        'HEARTBEAT'
    );
}
}

File: server-d/api/services/health_reporter.py
import logging

import requests
from django.conf import settings


logger = logging.getLogger(__name__)


class HealthReporter:

    def report(
        self,
        status,
        component,
        reason,
        message=None,
    ):
        payload = {
            "service": settings.HEALTH_SERVICE_NAME,
            "status": status,
            "component": component,
            "reason": reason,
            "message": message,
        }

        try:
            response = requests.post(
                f"{settings.AL_URL}/api/v1/health-events",
                json=payload,
                headers={
                    "Authorization": f"Bearer {settings.AL_HEALTH_TOKEN}",
                    "Accept": "application/json",
                },
                timeout=3,
            )

            if response.ok:
                logger.info(
                    "Health event sent to AL",
                    extra={
                        "payload": payload,
                        "response": response.json(),
                    },
                )

                return True

            logger.error(
                "AL rejected health event",
                extra={
                    "status_code": response.status_code,
                    "response": response.text,
                    "payload": payload,
                },
            )

            return False

        except requests.RequestException as exception:
            logger.error(
                "Unable to communicate with AL: %s",
                exception,
            )

            return False

    def healthy(
        self,
        component,
        reason="SERVICE_AVAILABLE",
    ):
        return self.report(
            "HEALTHY",
            component,
            reason,
        )

    def degraded(
        self,
        component,
        reason,
        message=None,
    ):
        return self.report(
            "DEGRADED",
            component,
            reason,
            message,
        )

    def down(
        self,
        component,
        reason,
        message=None,
    ):
        return self.report(
            "DOWN",
            component,
            reason,
            message,
        )

    def heartbeat(self):
        return self.report(
            "HEALTHY",
            "application",
            "HEARTBEAT",
        )